fling/utils/attack_utils/dlg.py

```diff
@@ -0,0 +1,184 @@
+        self.iteration = iteration
+        self.working_dir = working_dir
+        self.logger = Logger(self.working_dir)
+        self.base_index = 0
+        self.iteration_per_save = iteration_per_save
+
+    def attack(
+            self,
+            model: nn.Module,
+            dataset: Dataset,
+            device: str,
+            class_number: int,
+            parameter_args: Union[Dict, EasyDict] = {"name": "all"},
+            batch_size: bool = 1,
+            use_gt_labels: bool = True,
+            save_img: bool = False,
+            optim_backend: str = 'lbfgs',
+            tv_weight: float = 0,
+    ) -> float:
+        """
+        Overview:
+            The main attack function.
+        Arguments:
+            model: The model to be attacked.
+            dataset: The dataset tried to be recovered by the attacker.
+            batch_size: Batch-size of the input images. A larger batch size will be harder to attack. Default to be 1.
+            parameter_args: Parameters whose gradients are visible to attackers. Typically equals to \
+                ``aggregation_parameters`` in the config file.
+            device: Device to run this attack algorithm, such as ``"cpu"`` and ``"cuda"``.
+            class_number: The number of classes for the classification task.
+            use_gt_labels: Whether the ground truth label is able to be acquired by the attacker. Default to be \
+                ``True``. If set to be ``False``, it will be harder to attack.
+            save_img: Whether to save the reconstructed images into working dir. Default to be ``False``.
+            optim_backend: Backend for optimizing the recovred image. Supports "lbfgs" and "adam".
+            tv_weight: The weight of total variance loss (TV-loss). A larger weight can result in more "smooth" \
+                recovered images.
+        Returns:
+            mean_loss: The mean reconstruction loss across all samples in the dataset. A smaller value indicates a \
+                better performance for the attacker.
+        """
+        start, end = 0, batch_size
+        batch_idx = 0
+        total_loss = []
+
+        model = model.to(device=device)
+        model = model.eval()
+        tv_criterion = TVLoss()
+
+        # Get the parameters whose gradients are visible to the attacker.
+        if not isinstance(parameter_args, EasyDict):
+            parameter_args = EasyDict(parameter_args)
+        visible_parameters = get_weights(model, parameter_args)
+
+        # Iterate for each batch in the dataset.
+        while end <= len(dataset):
+            if start == end:
+                break
+            batch = dataset[start:end]
+            start = end
+            end = min(len(dataset), end + batch_size)
+            self.logger.logging(f'Star batch: {batch_idx}...')
+            batch_idx += 1
+
+            # Prepare input data and labels.
+            batch_x = torch.stack([batch[i][0] for i in range(len(batch))]).to(device)
```

Comment:
The data structure here is not consistent with the structure used in the training phase. It is recommended to be modified as `batch_x = torch.stack([batch[i]['input'] for i in range(len(batch))]).to(device)
            batch_y = torch.tensor([batch[i]['class_id'] for i in range(len(batch))]).to(device).unsqueeze(1)`